packages/transport-test/e2e/controller.ts

```diff
@@ -0,0 +1,173 @@
+        stopBridge: typeof TrezorUserEnvLinkClass.prototype.stopBridge;
+        startEmu: typeof TrezorUserEnvLinkClass.prototype.startEmu;
+        stopEmu: typeof TrezorUserEnvLinkClass.prototype.stopEmu;
+    };
+
+    constructor() {
+        super();
+
+        this.logger = console;
+
+        this.originalApi = {
+            connect: super.connect.bind(this),
+            startBridge: super.startBridge.bind(this),
+            stopBridge: super.stopBridge.bind(this),
+            startEmu: super.startEmu.bind(this),
+            stopEmu: super.stopEmu.bind(this),
+        };
+
+        this.connect = !env.USE_HW
+            ? this.originalApi.connect
+            : () => {
+                  return Promise.resolve(null);
+              };
+
+        this.startBridge =
+            !env.USE_HW && !env.USE_NODE_BRIDGE
+                ? (version?: string) => this.originalApi.startBridge(version)
+                : env.USE_NODE_BRIDGE
+                  ? async () => {
+                        this.nodeBridge = new TrezordNode({
+                            port: 21325,
+                            api: !env.USE_HW ? 'udp' : 'usb',
+                            logger: new Log('test-bridge', false),
+                        });
+
+                        await this.nodeBridge.start();
+
+                        // todo: this shouldn't be here, nodeBridge should be started when start resolves
+                        await this.waitForBridgeIsRunning(true);
+
+                        return null;
+                    }
+                  : () => this.waitForBridgeIsRunning(true);
+
+        this.stopBridge =
+            !env.USE_HW && !env.USE_NODE_BRIDGE
+                ? this.originalApi.stopBridge
+                : env.USE_NODE_BRIDGE
+                  ? async () => {
+                        await this.nodeBridge?.stop();
+
+                        // todo: this shouldn't be here, nodeBridge should be stopped when stop resolves
+                        await this.waitForBridgeIsRunning(false);
+
+                        return null;
+                    }
+                  : () => this.waitForBridgeIsRunning(false);
+
+        this.startEmu = !env.USE_HW
+            ? this.originalApi.startEmu
+            : env.USE_HW
+              ? () => this.waitForNumberOfDevices(1)
+              : () => {
+                    console.log('todo: ping local emu');
+
+                    return Promise.resolve(null);
+                };
+
+        this.stopEmu = !env.USE_HW
+            ? this.originalApi.stopEmu
+            : env.USE_HW
+              ? () => this.waitForNumberOfDevices(0)
+              : () => {
+                    console.log('todo: ping local emu');
+
+                    return Promise.resolve(null);
+                };
+    }
+
+    private waitForNumberOfDevices = (expected: number) => {
+        this.logger.log(`waiting for ${expected} device to be connected`);
```

Comment:
this.logger.log(
            `${env.USE_HW ? '[MANUAL ACTION REQUIRED]' : ''} waiting for ${expected} device to be connected`,
        );